pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -467,50 +487,57 @@ func (c *Controller) Stop() {
 
 // Services implements a service catalog operation
 func (c *Controller) Services() ([]*model.Service, error) {
-	var k8sNodes []*kubernetesNode
+	c.RLock()
+	out := make([]*model.Service, 0, len(c.servicesMap))
+	for _, svc := range c.servicesMap {
+		out = append(out, svc)
+	}
+	c.updateServiceExternalAddr(out)
```

Comment:
I think updating it here is less efficient than what id do. because Services() is more frequently called.

Reply:
Its not. It’s called only once by the push context initServices() code. Also if I tried the other way, and added full pushes to nodes, the code gets stuck. All tests fail!

Reply:
And not to mention we will need to have locks in place for the services. That is less efficient as well because we have to acquire mutex lock for each and every service.

Reply:
I know, but every other kind of configs will trigger this too.

Reply:
>And not to mention we will need to have locks in place for the services.

In a mesh, it is not needed to update this often. As once we deployed all the multicluster env, the gateway service or even the nodes are fixed. We donot need 

BTW, The locks of the service is really hacky, need to do a further refactor.